pkg/sources/github/feedlet.go

```diff
@@ -0,0 +1,396 @@
+		glog.Warningf("failed to create service: %v", err)
+		return nil, err
+	}
+
+	glog.Infof("created Service: %+v", service)
+
+	// Start watching the Receive Adapter Service for it's updated domain name. This will be passed
+	// to GitHub as part of the webhook registration.
+	receiveAdaptorDomain, err := t.waitForServiceDomain(service.GetObjectMeta().GetName())
+	if err != nil {
+		glog.Infof("failed to get the service: %v", err)
+	}
+
+	return t.createWebhook(trigger, receiveAdaptorDomain)
+}
+
+func (t *GithubEventSource) waitForServiceDomain(serviceName string) (string, error) {
+	sc := t.servingclientset.ServingV1alpha1().Services(t.feedNamespace)
+
+	wt := watchTimeout
+	opts := metav1.ListOptions{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Service",
+			APIVersion: "v1alpha1",
+		},
+		FieldSelector:  fmt.Sprintf("metadata.name=%s", serviceName),
+		LabelSelector:  "receive-adapter=github",
+		Watch:          true,
+		TimeoutSeconds: &wt,
+	}
+
+	w, err := sc.Watch(opts)
+	if err != nil {
+		glog.Infof("failed to create watch: %v", err)
+		return "", err
+	}
+
+	eventChan := w.ResultChan()
+	for {
+		event, more := <-eventChan
+		if !more {
+			glog.Infof("expected a Service object, but got %T", event.Object)
+			return "", fmt.Errorf("no more")
+		}
+		switch event.Type {
+		case watch.Error:
+			return "", fmt.Errorf("watched Service error")
+		case watch.Deleted:
+			return "", fmt.Errorf("watched Service deleted")
+		case watch.Added:
+			fallthrough
+		case watch.Modified:
+			service, ok := event.Object.(*v1alpha1.Service)
+			if !ok {
+				glog.Infof("expected a Service object, but got %T", event.Object)
+				continue
+			}
+			if service.Name == serviceName {
+				status := service.Status
+				if status.Domain != "" {
+					w.Stop()
+					glog.Infof("got domain: %s", event.Object)
+					return status.Domain, nil
+				}
+			}
+		}
+	}
+}
+
+func receiveAdapterName(resource string) string {
+	serviceName := fmt.Sprintf("%s-%s-%s", "github", resource, postfixReceiveAdapter) // TODO: this needs more UUID on the end of it.
+	serviceName = strings.Replace(serviceName, "/", "-", -1)
+	serviceName = strings.Replace(serviceName, ".", "-", -1)
+	return serviceName
+}
+
+func (t *GithubEventSource) createReceiveAdapter(trigger sources.EventTrigger, target string) (*v1alpha1.Service, error) {
+	sc := t.servingclientset.ServingV1alpha1().Services(t.feedNamespace)
+
+	serviceName := receiveAdapterName(trigger.Resource)
+
+	// First, check if service exists already.
+	if _, err := sc.Get(serviceName, metav1.GetOptions{}); err != nil {
+		if !apierrs.IsNotFound(err) {
+			glog.Infof("service.Get for %q failed: %s", serviceName, err)
+			return nil, err
+		}
+		glog.Infof("service %q doesn't exist, creating", serviceName)
+	} else {
+		glog.Infof("found existing service %q", serviceName)
+		return nil, nil
+	}
+
+	glog.Infof("secretName %s ; secretKey: %s", trigger.Parameters[secretName].(string), trigger.Parameters[secretKey].(string))
```

Comment:
Please don't log secrets/credentials.

Reply:
I'm actually vaguely surprised that K8S API core doesn't have some protections against this. E.g.

```
// Secret is a string that protects itself against accidental printing
// To use a secret, cast it into a string with string(mySecret). Do not
// ever store the result of this cast as a named variable.
type Secret string

func (Secret) String() string {
  return "### REDACTED ###"
}
```